perf(ngcc): add support for parallel execution of tasks

gkalpak opened the pull request:
This is a continuation of the work started in #32052 to speed up `ngcc` by adding support for executing tasks in parallel (when async execution is acceptable).

##
This PR includes numerous refactorings to clean up/simplify the code and prepare the introduction of  parallel task execution. It also includes a tiny fix: ``only back up the original `prepublishOnly` script and not the overwritten one``

See the individual commit messages for more details.

---

**Notes on testing**

I have added unit tests for all added/modified code except for the `ClusterMaster` class, because:
- It mostly orchestrates other parts (that are independently tested).
- The most important parts of its functionality are covered by integration tests.
- I didn't want to spend too much time on testing it before a first review (in case the implementation changed drastically).

The integration tests at [ngcc/test/integration/ngcc_spec.ts](https://github.com/angular/angular/blob/f5bec3ff5/packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts), which are run under bazel, don't play well with spawning multiple processes, so they are only run on one process (which means they don't exercise the `ClusterExecutor`).

The integration tests at [integration/ngcc/test.sh](https://github.com/angular/angular/blob/f5bec3ff5/integration/ngcc/test.sh) do run in parallel mode (and only in that mode, since that is the only option with `ivy-ngcc`).

Other than that, I have manually verified that `ngcc` can successfully build angular.io in parallel mode (with various command-line argument combinations) and that the resulting `node_modules/`are identical to those produced in sync mode with the following exceptions:
- `package.json > __processed_by_ivy_ngcc__` might have properties in different order.
    _This is expected, since now the order is which tests are run/completed is different and non-deterministic._
- Some generated variable names can be different (mostly(/only?) noticed in UMD bundles). E.g. `data_r50` vs `data_r450`.
    _Not sure why that is and whether it is problematic._

##
**Notes on performance**

For processing angular.io, I have observed speed improvement between 1.8x (on my Windows laptop - 7 workers) to 2.6x (on Linux CI - 8 workers). This was a little surprizing, since task processing (which runs on multiple processes) takes up 80%-90% and all workers are busy most of the time (i.e. there are rarely idle workers).

More investigation is needed to better understand/improve those numbers, but this can happen after the PR lands.

Fun fact:
There doesn't seem to be much improvement above 4 workers. Even with up to 20 workers on CI, the duration was the same (if not worse: 29s vs 23s).

---

**TODO**

- [x] Get the PR reviewed.
- [x] Determine whether the differences in variable names (in UMD bundles?) are an issue.<sup>1</sup>
- [x] ~~Determine whether we need unit tests for `ClusterMaster`.~~

<sup>1</sup>: The differences are in "shared context vars" generated via [BindingScope#generateSharedContextVar()][1] by the [TemplateDefinitionBuilder][2] instantiated in [compileComponentFromMetadata()][3]. Since these are local to the template, we don't mind having duplicate names in different entry-points or formats.

[1]: https://github.com/angular/angular/blob/4c3674f66/packages/compiler/src/render3/view/template.ts#L1651-L1652
[2]: https://github.com/angular/angular/blob/4c3674f66/packages/compiler/src/render3/view/compiler.ts#L229-L232
[3]: https://github.com/angular/angular/blob/4c3674f66/packages/compiler/src/render3/view/compiler.ts#L187

gkalpak committed:
refactor(ngcc): minor code clean-up following #32052

This commit addresses the review feedback from #32052, which was merged
before addressing the feedback there.

gkalpak committed:
fix(ngcc): only back up the original `prepublishOnly` script and not the overwritten one

In order to prevent `ngcc`'d packages (e.g. libraries) from getting
accidentally published, `ngcc` overwrites the `prepublishOnly` npm
script to log a warning and exit with an error. In case we want to
restore the original script (e.g. "undo" `ngcc` processing), we keep a
backup of the original `prepublishOnly` script.

Previously, running `ngcc` a second time (e.g. for a different format)
would create a backup of the overwritten `prepublishOnly` script (if
there was originally no `prepublishOnly` script). As a result, if we
ever tried to "undo" `ngcc` processing and restore the original
`prepublishOnly` script, the error-throwing script would be restored
instead.

This commit fixes it by ensuring that we only back up a `prepublishOnly`
script, iff it is not the one we created ourselves (i.e. the
error-throwing one).

gkalpak committed:
refactor(ngcc): abstract updating `package.json` files behind an interface

To persist some of its state, `ngcc` needs to update `package.json`
files (both in memory and on disk).

This refactoring abstracts these operations behind the
`PackageJsonUpdater` interface, making it easier to orchestrate them
from different contexts (e.g. when running tasks in parallel on multiple
processes).

Inspired by/Based on @alxhub's prototype: alxhub/angular@cb631bdb1

gkalpak committed:
refactor(ngcc): abstract work orchestration/execution behind an interface

This change does not alter the current behavior, but makes it easier to
introduce new types of `Executors` , for example to do the required work
in parallel (on multiple processes).

Inspired by/Based on @alxhub's prototype: alxhub/angular@cb631bdb1

gkalpak committed:
refactor(ngcc): add support for asynchronous execution

Previously, `ngcc`'s programmatic API would run and complete
synchronously. This was necessary for specific usecases (such as how the
`@angular/cli` invokes `ngcc` as part of the TypeScript module
resolution process), but not for others (e.g. running `ivy-ngcc` as a
`postinstall` script).

This commit adds a new option (`async`) that enables turning on
asynchronous execution. I.e. it signals that the caller is OK with the
function call to complete asynchronously, which allows `ngcc` to
potentially run in a more efficient mode.

Currently, there is no difference in the way tasks are executed in sync
vs async mode, but this change sets the ground for adding new execution
options (that require asynchronous operation), such as processing tasks
in parallel on multiple processes.

NOTE:
When using the programmatic API, the default value for `async` is
`false`, thus retaining backwards compatibility.
When running `ngcc` from the command line (i.e. via the `ivy-ngcc`
script), it runs in async mode (to be able to take advantage of future
optimizations), but that is transparent to the caller.

gkalpak committed:
refactor(ngcc): abstract task selection behind an interface

This change does not alter the current behavior, but makes it easier to
introduce `TaskQueue`s implementing different task selection algorithms,
for example to support executing multiple tasks in parallel (while
respecting interdependencies between them).

Inspired by/Based on @alxhub's prototype: alxhub/angular@cb631bdb1

gkalpak committed:
refactor(ngcc): implement task selection for parallel task execution

This commit adds a new `TaskQueue` implementation that supports
executing multiple tasks in parallel (while respecting interdependencies
between them).

This new implementation is currently not used, thus the behavior of
`ngcc` is not affected by this change. The parallel `TaskQueue` will be
used in a subsequent commit that will introduce parallel task execution.

gkalpak committed:
refactor(ngcc): take advantage of early knowledge about format property processability

In the past, a task's processability didn't use to be known in advance.
It was possible that a task would be created and added to the queue
during the analysis phase and then later (during the compilation phase)
it would be found out that the task (i.e. the associated format
property) was not processable.

As a result, certain checks had to be delayed, until a task's processing
had started or even until all tasks had been processed. Examples of
checks that had to be delayed are:
- Whether a task can be skipped due to `compileAllFormats: false`.
- Whether there were entry-points for which no format at all was
  successfully processed.

It turns out that (as made clear by the refactoring in 9537b2ff8), once
a task starts being processed it is expected to either complete
successfully (with the associated format being processed) or throw an
error (in which case the process will exit). In other words, a task's
processability is known in advance.

This commit takes advantage of this fact by moving certain checks
earlier in the process (e.g. in the analysis phase instead of the
compilation phase), which in turn allows avoiding some unnecessary work.
More specifically:

- When `compileAllFormats` is `false`, tasks are created _only_ for the
  first suitable format property for each entry-point, since the rest of
  the tasks would have been skipped during the compilation phase anyway.
  This has the following advantages:
  1. It avoids the slight overhead of generating extraneous tasks and
     then starting to process them (before realizing they should be
     skipped).
  2. In a potential future parallel execution mode, unnecessary tasks
     might start being processed at the same time as the first (useful)
     task, even if their output would be later discarded, wasting
     resources. Alternatively, extra logic would have to be added to
     prevent this from happening. The change in this commit avoids these
     issues.
- When an entry-point is not processable, an error will be thrown
  upfront without having to wait for other tasks to be processed before
  failing.

gkalpak committed:
refactor(ngcc): remove unused `EntryPointProcessingMetadata` data and types

Previously, `ngcc` needed to store some metadata related to the
processing of each entry-point. This metadata was stored in a `Map`, in
the form of `EntryPointProcessingMetadata` and passed around as needed.

After some recent refactorings, it turns out that this metadata (with
its only remaining property, `hasProcessedTypings`) was no longer used,
because the relevant information was extracted from other sources (such
as the `processDts` property on `Task`s).

This commit cleans up the code by removing the unused code and types.

gkalpak committed:
perf(ngcc): process tasks in parallel in async mode

`ngcc` supports both synchronous and asynchronous execution. The default
mode when using `ngcc` programmatically (which is how `@angular/cli` is
using it) is synchronous. When running `ngcc` from the command line
(i.e. via the `ivy-ngcc` script), it runs in async mode.

Previously, the work would be executed in the same way in both modes.

This commit improves the performance of `ngcc` in async mode by
processing tasks in parallel on multiple processes. It uses the Node.js
built-in [`cluster` module](https://nodejs.org/api/cluster.html) to
launch a cluster of Node.js processes and take advantage of multi-core
systems.

Preliminary comparisons indicate a 1.8x to 2.6x speed improvement when
processing the angular.io app (apparently depending on the OS, number of
available cores, system load, etc.). Further investigation is needed to
better understand these numbers and identify potential areas of
improvement.

Inspired by/Based on @alxhub's prototype: alxhub/angular@cb631bdb1
Original design doc: https://hackmd.io/uYG9CJrFQZ-6FtKqpnYJAA?view

Jira issue: [FW-1460](https://angular-team.atlassian.net/browse/FW-1460)

gkalpak committed:
refactor(ngcc): add debug logging for the duration of different operations

This gives an overview of how much time is spent in each operation/phase
and makes it easy to do rough comparisons of how different
configurations or changes affect performance.

mary-poppins commented:
You can preview 42b152c at https://pr32427-42b152c.ngbuilds.io/.

hiepxanh commented:
do it bro, we need ngc be faster and multi thread. every second you saving even `3 seconds` is 1000 developers * 3 = 50 hours everytime they build their app. You are doing something is bigger meaning than you thought.

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -115,8 +126,18 @@ export function mainNgcc(
+  //       of `PackageJsonUpdater` that actually writes to disk (across all processes).
+  //       This is hard to enforce automatically, when running on multiple processes, so needs to be
+  //       enforced manually.
+  const pkgJsonUpdater = getPackageJsonUpdater(inParallel, fileSystem);
```
I believe that this could move inside the `Executor`. That way we can actually create a different updater for master and slave, rather than having an intelligent one that needs to care about the `cluster` API.

petebacondarwin replied:
In fact I think that having a `ClusterMasterExecutor` and `ClusterWorkerExecuter` would be a cleaner way to architect this. I think it would require moving a lot of the helper functions in the file into an abstract `BaseExecutor` class, but I think this would also be helpful in terms of making this big main file more manageable.

gkalpak replied:
I will add this to the list of things to consider/improve/try for future PRs.

petebacondarwin reviewed (approved):
Nothing here that I think should block the merging of this PR.
A handful of nits that I expect you can clean up before merge.
And a fairly chunky architectural suggestion, which if agreed could be done in a subsequent refactoring PR.

gkalpak commented on packages/compiler-cli/ngcc/src/writing/package_json_updater.ts:
```diff
@@ -0,0 +1,160 @@
+    // NOTE: We are not using `preExistingParsedJson` (even if specified) to avoid corrupting the
+    //       content on disk in case `preExistingParsedJson` is outdated.
+    const parsedJson =
+        this.fs.exists(packageJsonPath) ? JSON.parse(this.fs.readFile(packageJsonPath)) : {};
```
The problem is that the in-memory one is a result of the combination of the on-disk with any overrides defined in an `ngcc` config file. So, they aren't guaranteed to be identical.

##
Also, theoretically, we only need to ensure that any changes go through the `PackageJsonUpdater`, which will keep them "in sync" (as much in sync as they need to be, but not necessarily identical).

BTW, I don't think that updating the in-memory one currently matters (i.e. we don't need to read the updates on the same `ngcc` invocation), but I somehow felt uneasy deliberately allowing them to get out-of-sync 👼

petebacondarwin replied:
If we require that all changes should be through a `PackageJsonUpdater` then these should always be in sync, right? In which case we could optimize to just update the in-memory object, if available, and then write that to the disk.

gkalpak replied:
They don't have to be the same:

> The problem is that the in-memory one is a result of the combination of the on-disk with any overrides defined in an `ngcc` config file. So, they aren't guaranteed to be identical.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/api.ts:
```diff
@@ -20,11 +20,21 @@ export type CompileFn = (task: Task) => void;
+export interface Executor {
+  execute(
+      analyzeFn: AnalyzeEntryPointsFn, createCompileFn: CreateCompileFn,
+      options: ExecutionOptions): void;
```
Resolved (see https://github.com/angular/angular/pull/32427#discussion_r319718630).

gkalpak committed:
fixup! refactor(ngcc): abstract task selection behind an interface

gkalpak commented on packages/compiler-cli/ngcc/test/execution/task_selection/parallel_task_queue_spec.ts:
```diff
@@ -0,0 +1,556 @@
+
+         expect(() => new ParallelTaskQueue(tasks, graph))
+             .toThrowError(
+                 'Invariant violated: Multiple tasks are assigned generating typings for ' +
```
I was inspired by `service-worker` code (from @alxhub) :grin: https://github.com/angular/angular/blob/04d4fea3e8784ea81916b24f3b153331967bad71/packages/service-worker/worker/src/driver.ts#L617

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/api.ts:
```diff
@@ -0,0 +1,49 @@
+}
+
+/** The type of messages sent from the cluster master to cluster workers. */
+export type MessageFromMaster = ProcessTaskMessage;
```
I don't have a string preference.
Would you keep `MessageFromWorker` or change that one to `MessageToMaster` too?

petebacondarwin replied:
My feeling was that messages would go to and from something. So `MessageFromWorker` and `MessageToWorker`.

gkalpak replied:
Well, these messages always go _from_ something _to_ something. I chose to name the origin in both cases (to avoid the verbose `MessageFromMasterToWorker` and `MessageFromWorkerToMaster` 😛).

Changing them to be relative to workers...

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/master.ts:
```diff
@@ -0,0 +1,258 @@
+      this.logger.debug(
+          `All ${this.taskAssignments.size} workers are currently busy and cannot take on more ` +
+          'work.');
+    } else if (!isWorkAvailable) {
```
Good catch 👍
Removing it...

gkalpak committed:
fixup! perf(ngcc): process tasks in parallel in async mode

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -32,6 +40,7 @@ import {InPlaceFileWriter} from './writing/in_place_file_writer';
 import {NewEntryPointFileWriter} from './writing/new_entry_point_file_writer';
 import {DirectPackageJsonUpdater, PackageJsonUpdater} from './writing/package_json_updater';
 
+
```
Welll...theoretically most whitespace is unnecessary 😛

##
**Background:**
I don't understand what clang-format's preference is wrt to newlines around imports, but it seems to often (above a certain number of import statements?) like separate "absolute" imports, `../` imports and `./` imports into groups and add a newline between each group. So, I opted to do this manually (if any group had more than one entry).

When "import groups" are separated like that, having two newlines between imports and the rest of the code makes it easier to visually detect where imports end and the "interesting" code begins.

There doesn't seem to be a definite convention around that (I've seen anywhere between 1 and 3 newlines separating imports from the rest of the code), although admittedly one-line separators are more frequent.

##
So, the bottom line is:
The double-newlines were intentional, but I'm OK removing them, if you don't like them :grin:

petebacondarwin replied:
My experience of the clang formatter is this (disclaimer I am only guessing here):

* if it doesn't need to reorder anything in the imports list then it doesn't touch them at all, so you can add and delete empty lines as you like
* if it needs to reorder the imports then it will group them by level (absolute/package imports at the top, then `../../` then `../`, then './` last.

In order to do this grouping it naively adds a blank line after each group - not taking into account whether there was already a blank line there. So you sometimes end up with two blank lines.

So I don't believe that these double blank lines are intentional.

gkalpak replied:
No, the double blank lines were (intentionally) added by me, not the formatter :grin:

petebacondarwin replied:
Oh. OK...

gkalpak replied:
So, can I keep my double blank lines or would you prefer to see them gone? 🙈

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -32,48 +32,73 @@ import {NewEntryPointFileWriter} from './writing/new_entry_point_file_writer';
+/**
+ * The options to configure the ngcc compiler.
+ */
+export type NgccOptions = AsyncNgccOptions | SyncNgccOptions;
```
I couldn't find a way to do that and at the same time:
1. Keep TS happy wrt [process()][1].
2. Do not break the public API (not that [NgccOptions][2] and [process()][1] are public API).
3. Have TS show the correct docs.

For example see [this example][3].
If you have ideas, please share. 

[1]: https://github.com/angular/angular/blob/42b152c0c32cc1f7fd8081f26f4488add7972b0d/packages/compiler-cli/ngcc/index.ts#L16-L22
[2]: https://github.com/angular/angular/blob/42b152c0c32cc1f7fd8081f26f4488add7972b0d/packages/compiler-cli/ngcc/index.ts#L10
[3]: http://www.typescriptlang.org/play/#code/PQKhAIBUAsFNwPYAcAuBLBA7AzuFDwBjLAMzQHMBXAJ3hTnE3MMKIQFsk0AbWagOnAhgAWABQsAB5IE1FODSYUfEgENC8AHLNCAeVQYc4AN7jw4AEarssAAqr6ALnDYU1ReQDcZtpx6wAQW5uADFZdgdsAH5nCwQEXlVMbzFzUAgAEQRCXBJZcAADawBPTEIC5wBJFAByXAK3SlgCxGpCtW4bAsFhHxKymMt4xOTxAF9xcXSoBmR0LFx8NkwyKlo8BiYWXy5eNry2-sJoaiwESlwpWEJKecwe0QlpWXkUYqR4AOxSwm0WfTuuAAvOBdOw0CgADx-PQGBYAGnANSONQAfOAAGQmI7ORqwMYpKZgGbwOaGRYEYgrCg0OibHQ7fz7fLfMonM4XcBXG53B7iKQyOR4d7wADKPxhAPJ4BBYIh0J0UoRSJR6Kxxhx4A6NgJkyegvkJEoZTu4AiigAFHNsM4vhLFXCcABKZy2U7gmyQgBuCDQABNUSkBS8tcbCKbzZgragbeBxWVJY7sC7wD7-UHnkKjSbDGbVJbrc5E4CU2m-eAAD7gN0cNCesvo0xiCZifmZw1h01IU4abDYaMoWN2hMOkuu9112De30BjMG0M5rDgbvZWB9gex+O-Ufk0szuch7Ph3Mr3v9wvgYu75xlyvVif1meNnzmWgoGiYPMFmNOlItoA

petebacondarwin replied:
I think this would do it ([playground](http://www.typescriptlang.org/play/#code/PQKhAIBUAsFNwPYAcAuBLBA7AzuFDwBjLAMzQHMBXAJ3hTnE3MMKIQFsk0AbWagOnAhgAWABQsAB5IE1FODSYUfEgENC8AHLNCAeVQYc4AN7jw4AEarssAAqr6ALnDYU1ReQDcZtpx6wAQW5uADFZdgdsAH5nCwQEXlVMbzFzUAgAEQRCXBJZcAADawBPTEIC5wBJFAByXAK3SlgCxGpCtW4bAsFhHxKymMt4xOTxAF9xcXSoBmR0LFx8NkwyKlo8BiYWXy5eNry2-sJoaiwESlwpWEJKecwe0QlpWXkUYqR4AOxSwm0WfTuuAAvOBdOw0CgADx-PQGBYAGnANSONQAfOAAGQmI7ORqwMYpKZgGbwOaGRYEYgrCg0OibHQ7fz7fLfMonM4XcBXG53B7iKQyOR4d7wADKPxhAPJ4BBYIh0J0UoRSJR6Kxxhx4A6NgJkyegvkJEoZTu4AiigAFHNsM4vhLFXCcABKZy2U7gmyQgBuCDQABNUSkBS8tcbCKbzZgragbeBxWVJY7sC7wD7-UHnkKjSbDGbVJbrc5E4CU2m-eAAD7gN0cNCesuB-mZw1hiP5qOF8DF8ml33lqs1j2wb199GmMQTMRNg2hnNYcBIU4abDYaMoWN2hMOkuu9114cNjMz7Ph3OL7KwFdr2Px37bnvOMtHkMn03n5erzvdha9-2V6t7vWo7PlmrZnkul6fjGRb3j+j59v+g77iO-pjj45i0CgNCYHmBYxk6KSTkAA)) but perhaps this is not acceptable?

```ts
export function main(opts: AsyncNgccOptions): Promise<void>;
export function main(opts: SyncNgccOptions): void;
export function main(opts: NgccOptions): void | Promise<void>;
export function main(opts: NgccOptions): void | Promise<void> {
}
```

petebacondarwin replied:
Although to be honest I don't think there is much benefit in overloading the `process` function like this. Just having a single syntax would be enough, right?

```ts
export function process(opts: NgccOptions): void | Promise<void> {
```

Unless the Async and Sync versions of the type will have other properties there is not much benefit regarding this function signature.

gkalpak replied:
The problem with https://github.com/angular/angular/pull/32427#discussion_r320738999 and the reason we need the overloaded signatures is that it allows TS to infer the return type based on the options passed.

If the options have `async: true`, then it knows the returned value is a promise. Otherwise, it knows it is void. This also retains backwards compatibility for the cli (and any other consumer of `process()`).

petebacondarwin replied:
How about something like this:

```diff
diff --git a/packages/compiler-cli/ngcc/index.ts b/packages/compiler-cli/ngcc/index.ts
index 90d58bb7d1..916a2249cb 100644
--- a/packages/compiler-cli/ngcc/index.ts
+++ b/packages/compiler-cli/ngcc/index.ts
@@ -7,15 +7,19 @@
  */
 import {CachedFileSystem, NodeJSFileSystem, setFileSystem} from '../src/ngtsc/file_system';
 
-import {AsyncNgccOptions, NgccOptions, SyncNgccOptions, mainNgcc} from './src/main';
+import {AsyncNgccOptions, SyncNgccOptions, mainNgcc} from './src/main';
 export {ConsoleLogger, LogLevel} from './src/logging/console_logger';
 export {Logger} from './src/logging/logger';
-export {AsyncNgccOptions, NgccOptions, SyncNgccOptions} from './src/main';
+export {AsyncNgccOptions, SyncNgccOptions} from './src/main';
 export {PathMappings} from './src/utils';
 
-export function process(options: AsyncNgccOptions): Promise<void>;
-export function process(options: SyncNgccOptions): void;
-export function process(options: NgccOptions): void|Promise<void> {
+export function process(options: SyncNgccOptions): void|Promise<void> {
+  // Recreate the file system on each call to reset the cache
+  setFileSystem(new CachedFileSystem(new NodeJSFileSystem()));
+  return mainNgcc(options);
+}
+
+export function processAsync(options: AsyncNgccOptions): void|Promise<void> {
   // Recreate the file system on each call to reset the cache
   setFileSystem(new CachedFileSystem(new NodeJSFileSystem()));
   return mainNgcc(options);
diff --git a/packages/compiler-cli/ngcc/src/main.ts b/packages/compiler-cli/ngcc/src/main.ts
index 4e7b7dc3df..fcd4ba370e 100644
--- a/packages/compiler-cli/ngcc/src/main.ts
+++ b/packages/compiler-cli/ngcc/src/main.ts
@@ -42,9 +42,9 @@ import {DirectPackageJsonUpdater, PackageJsonUpdater} from './writing/package_js
 
 
 /**
- * The options to configure the ngcc compiler for synchronous execution.
+ * The options to configure the ngcc compiler.
  */
-export interface SyncNgccOptions {
+export interface NgccOptions {
   /** The absolute path to the `node_modules` folder that contains the packages to process. */
   basePath: string;
 
@@ -89,25 +89,34 @@ export interface SyncNgccOptions {
    * Provide a file-system service that will be used by ngcc for all file interactions.
    */
   fileSystem?: FileSystem;
+}
 
+/**
+ * The options to configure the ngcc compiler for asynchronous execution.
+ */
+export interface AsyncNgccOptions extends NgccOptions
+{
   /**
    * Whether the compilation should run and return asynchronously. Allowing asynchronous execution
    * may speed up the compilation by utilizing multiple CPU cores (if available).
    *
    * Default: `false` (i.e. run synchronously)
    */
-  async?: false;
-}
-
-/**
- * The options to configure the ngcc compiler for asynchronous execution.
- */
-export type AsyncNgccOptions = Omit<SyncNgccOptions, 'async'>& {async: true};
+  async: true;
+};
 
 /**
- * The options to configure the ngcc compiler.
+ * The options to configure the ngcc compiler for synchronous execution.
  */
-export type NgccOptions = AsyncNgccOptions | SyncNgccOptions;
+export interface SyncNgccOptions extends NgccOptions {
+  /**
+   * Whether the compilation should run and return asynchronously. Allowing asynchronous execution
+   * may speed up the compilation by utilizing multiple CPU cores (if available).
+   *
+   * Default: `false` (i.e. run synchronously)
+   */
+  async?: false;
+}
 
 const EMPTY_GRAPH = new DepGraph<EntryPoint>();
 
@@ -119,12 +128,12 @@ const EMPTY_GRAPH = new DepGraph<EntryPoint>();
  *
  * @param options The options telling ngcc what to compile and how.
  */
-export function mainNgcc(options: AsyncNgccOptions): Promise<void>;
 export function mainNgcc(options: SyncNgccOptions): void;
+export function mainNgcc(options: AsyncNgccOptions): Promise<void>;
 export function mainNgcc(
     {basePath, targetEntryPointPath, propertiesToConsider = SUPPORTED_FORMAT_PROPERTIES,
      compileAllFormats = true, createNewEntryPointFormats = false,
-     logger = new ConsoleLogger(LogLevel.info), pathMappings, async = false}: NgccOptions): void|
+     logger = new ConsoleLogger(LogLevel.info), pathMappings, async = false}: AsyncNgccOptions|SyncNgccOptions): void|
     Promise<void> {
   // Execute in parallel, if async execution is acceptable and there are more than 1 CPU cores.
   const inParallel = async && (os.cpus().length > 1);
```

gkalpak replied:
`NgccOptions` is part of the public API, so removing it would be breaking change.
In the interest of moving this PR forward, I will leave things as is. (Feel free to suggest changes in follow-up PRs.)

gkalpak commented on packages/compiler-cli/ngcc/src/execution/single_process_executor.ts:
```diff
@@ -22,13 +22,15 @@ export class SingleProcessExecutor implements Executor {
     this.logger.debug(`Running ngcc on ${this.constructor.name}.`);
 
-    const tasks = analyzeFn();
+    const taskQueue = analyzeFn();
```
I am fine with either (I think both are true - it both analyzes the entry points and computes the tasks :grin:).

In previous stages, it returned more than just a `TaskQueue` (e.g. the `DepGraph`), so I would have argued that "analyze" was more accurate. Now that it only returns the `TaskQueue`, I am fine with "compute tasks" tasks. (I probably still slightly prefer the term "analyze", as I find it not unlikely that it might do more than compute the tests in the future :wink:)

WDYT?

petebacondarwin replied:
OK, so how about `analyzeEntryPointsFn`?
Or even better (IMO) `analyzeEntryPoints` - adding `Fn` seems fine when you are passing the function around but when you are calling it, it feels like `analyzeFn()` sounds like you are analyzing the `fn` :-)

gkalpak replied:
Works for me 👍

gkalpak commented on packages/compiler-cli/ngcc/src/execution/task_selection/base_task_queue.ts:
```diff
@@ -0,0 +1,51 @@
+  }
+  protected inProgressTasks = new Set<Task>();
+
+  constructor(protected tasks: Task[]) {}
```
The `TaskQueue` is already taking advantage of the fact that the `Task[]` list is "ordered", but I agree that having a more descriptive type would make it more obvious in the code (instead of relying solely on comments).

gkalpak committed:
fixup! refactor(ngcc): minor code clean-up following #32052

gkalpak commented:
Rebased on master and addressed feedback (either by adding fixup commits or by commenting).
@petebacondarwin, ptal.

I, also, created #32486 to capture ideas for future improvements/refactorings (as discussed here and on Slack).

mary-poppins commented:
You can preview 0db6569 at https://pr32427-0db6569.ngbuilds.io/.

IgorMinar reviewed (approved):
reviewed the integration tests part and skimmed through the rest. one nit, otherwise lgtm

IgorMinar commented:
thanks George for implementing this feature!

gkalpak commented:
For reference:
Igor's nit was that we should give better error messages in `integration/ngcc/test.sh` were we `exit 1`.

gkalpak committed:
test(ngcc): print an informative message when an integration test fails

mary-poppins commented:
You can preview ecb794c at https://pr32427-ecb794c.ngbuilds.io/.

mary-poppins commented:
You can preview 077a637 at https://pr32427-077a637.ngbuilds.io/.

mary-poppins commented:
You can preview c9f5903 at https://pr32427-c9f5903.ngbuilds.io/.
